### PR TITLE
[Clang][Parser] Have the depth of the abbreviated generic lambdas inside a requires clause differ from the surrounding generic lambda

### LINKED ISSUE #78524: Stack dump bug
I apologize in advance that my example is large, but I could not reproduce it in a smaller size. I tried to pass arguments using a nested lambda function to a require constraint but got an error. I also wrote a definition directive that switches between a concept and a lambda function (equally constrained), and got different results.

https://godbolt.org/z/41zfjdoTs

### PR DESCRIPTION
**Pull request by zyn0217**
A one-line fix, again : )

This fixes https://github.com/llvm/llvm-project/issues/78524 and the similar example at https://github.com/llvm/llvm-project/issues/78524#issuecomment-1899886951.

We previously increased the template depth by one after parsing the attaching requires-clause on a lambda expression. This led to a problem where the 'auto' parameters of nested abbreviated generic lambdas, inside of a requires-expression, had the same depth as the template parameters of the surrounding lambda. Consequently, during the concept-checking stage, we ended up substituting these parameters with the wrong template arguments because they were at different levels.

### CODE REVIEW

File: clang/lib/Parse/ParseExprCXX.cpp [hunk 1]
```diff
@@ -1385,6 +1385,11 @@ ExprResult Parser::ParseLambdaExpressionAfterIntroducer(
       Diag(RAngleLoc,
            diag::err_lambda_template_parameter_list_empty);
     } else {
+      // We increase the template depth before recursing into a requires-clause.
+      // The abbreviated generic lambdas thereof could have different template
+      // depths, avoiding substituting into wrong template parameters during the
```
Comment (shafik) [inline clang/lib/Parse/ParseExprCXX.cpp:1390]: Can you point to where the use of `TemplateParameterDepth` happens? This makes sense but I am curious where this happens in this case.

File: clang/test/Parser/cxx-concepts-requires-clause.cpp [hunk 2]
```diff
@@ -168,3 +168,30 @@ auto lambda4 = [] requires(sizeof(char) == 1){}; // expected-error {{expected bo
...
+  decltype(T()(C<T>)) Type;
+};
+
+template <typename T, typename U> D<T> G(T, U) { return {}; }
+
+struct E {};
+
+void F() {
+  G([]<typename T>
+//     ~~~~~~~~~~ T: 0,0
```
Comment (shafik) [inline clang/test/Parser/cxx-concepts-requires-clause.cpp:188]: In these context of this PR these comments about depth make sense but for someone coming to this test outside of this context they won't. Can you add more "depth" to the comment :-)

**Review by shafik — COMMENTED** [review]
I think this makes sense but I would like @cor3ntin to approve.

File: clang/test/Parser/cxx-concepts-requires-clause.cpp (hunk 2 above)
Comment (zyn0217) [reply clang/test/Parser/cxx-concepts-requires-clause.cpp]: Oh, I was going to detail the comment after putting up the PR, but for some reason I forgot it. Added.

File: clang/lib/Parse/ParseExprCXX.cpp (hunk 1 above)
Comment (zyn0217) [reply clang/lib/Parse/ParseExprCXX.cpp]: Done.

**Comment by zyn0217** [comment]
@cor3ntin gently ping

**Comment by zyn0217** [comment]
@cor3ntin would you mind taking a look? thanks!

File: clang/lib/Parse/ParseExprCXX.cpp [hunk 3]
```diff
@@ -1385,6 +1385,16 @@ ExprResult Parser::ParseLambdaExpressionAfterIntroducer(
       Diag(RAngleLoc,
            diag::err_lambda_template_parameter_list_empty);
     } else {
+      // We increase the template depth before recursing into a requires-clause.
+      //
+      // This depth is used for setting up a LambdaScopeInfo (at
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseExprCXX.cpp:1390]:
```suggestion
      // This depth is used for setting up a LambdaScopeInfo (in
```

File: clang/docs/ReleaseNotes.rst [hunk 4]
```diff
@@ -193,6 +193,10 @@ Bug Fixes to C++ Support
 - Fixed an out-of-bounds error caused by building a recovery expression for ill-formed
   function calls while substituting into constraints.
   (`#58548 <https://github.com/llvm/llvm-project/issues/58548>`_)
+- Fixed an issue where template parameters of the nested abbreviated generic lambda within
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:196]:
```suggestion
- Fixed an issue where template parameters of a nested abbreviated generic lambda within
```

File: clang/docs/ReleaseNotes.rst [hunk 5]
```diff
@@ -193,6 +193,10 @@ Bug Fixes to C++ Support
...
+  a requires-clause lie at the same depth as those of the surrounding lambda. This,
+  in turn, results in the wrong template argument substitution during the constraint checking.
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:198]:
```suggestion
  in turn, results in the wrong template argument substitution during constraint checking.
```

File: clang/lib/Parse/ParseExprCXX.cpp [hunk 6]
```diff
@@ -1385,6 +1385,16 @@ ExprResult Parser::ParseLambdaExpressionAfterIntroducer(
...
+      // Sema::RecordParsingTemplateParameterDepth), which participates in the
+      // invented template parameters later at InventTemplateParameter.
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseExprCXX.cpp:1392]:
```suggestion
      // Sema::RecordParsingTemplateParameterDepth), which is used later when inventing 
      // template parameters in InventTemplateParameter.
```

File: clang/lib/Parse/ParseExprCXX.cpp [hunk 7]
```diff
@@ -1385,6 +1385,16 @@ ExprResult Parser::ParseLambdaExpressionAfterIntroducer(
...
+      //
+      // This way, abbreviated generic lambdas could have different template
+      // depths, avoiding substitution into wrong template parameters during the
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseExprCXX.cpp:1395]:
```suggestion
      // depths, avoiding substitution into the wrong template parameters during the
```

Comment (cor3ntin) [inline clang/lib/Parse/ParseExprCXX.cpp:1395]:
```suggestion
      // depths, avoiding substitution into wrong template parameters during constraints
```

**Review by cor3ntin — APPROVED** [review]
I have some suggestion to improve the (very useful) comments.

Otherwise, LGTM

**Comment by zyn0217** [comment]
All done. Thanks for the review & wording suggestions!
(I'll try the windows CI again before landing it.)

**Pull request merged**